lib/server/middleware/maintenance.js

```diff
@@ -0,0 +1,51 @@
+import wildcard from 'node-wildcard';
+import boolean from 'boolean';
+import SettingModel from '../models/setting';
+import PageModel from '../models/page';
+import parseSettings from 'helpers/parse-settings';
+
+const ignoreMaintenanceRedirect = (url, options) => {
+  let result = false;
+  let i = 0;
+  const patterns = ['/admin*', '/images*', '/fonts*', '/graphql*', '/assets*', options.maintenancePage];
+  for (i = 0; i < patterns.length; i++) {
+    result = wildcard(url, patterns[i]);
+    if (result) {
+      break;
+    }
+  }
+  return result;
+};
+
+export default async (req, res, next) => {
+  const settingsArr = await SettingModel
+    .find({
+      _id: {$in: ['maintenanceMode', 'useDefaultMaintenancePage', 'maintenancePage']}
+    })
+    .exec();
+  const settings = parseSettings(settingsArr);
+  const defaultMaintenancePage = '/pub-res/maintenance';
+  let maintenancePage = defaultMaintenancePage;
+  settings.useDefaultMaintenancePage = boolean(settings.useDefaultMaintenancePage);
+  settings.maintenanceMode = boolean(settings.maintenanceMode);
+
+  if (settings.maintenancePage && !settings.useDefaultMaintenancePage) {
+    const page = await PageModel
+      .findById(settings.maintenancePage)
+      .exec();
+    maintenancePage = (!page) ? maintenancePage : `/${page.slug}`;
+  }
+
+  if (!settings.maintenanceMode &&
+    (req.url.toLowerCase() === maintenancePage || req.url.toLowerCase() === defaultMaintenancePage)) {
+    return res.redirect('/');
+  }
+
+  if (ignoreMaintenanceRedirect(req.url, {maintenancePage}) ||
+      req.user ||
+      !settings.maintenanceMode) {
+    return next();
+  }
+
+  res.redirect(maintenancePage);
+};
```

Comment:
There is no need for `node-wildcard` and an ignore testing function. You can simply verify if the user is logged in or not. If not verify if the site is on maintance mode, in that case we should output the page right away and not let anything pass from here onwards. Otherwise the graphQL API will be always accessible.